File: solutions/0056-merge-intervals/solution.py
from typing import List
import heapq

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []

        heap = [(s, e) for s, e in intervals]
        heapq.heapify(heap)

        merged = []
        s, e = heapq.heappop(heap)

        while heap:
            ns, ne = heapq.heappop(heap)
            if ns <= e:
                e = max(e, ne)
            else:
                merged.append([s, e])
                s, e = ns, ne

        merged.append([s, e])
        return merged

File: solutions/0056-merge-intervals/solution.java
import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {
        // Step 1: Intervals ko unke start time ke basis pe sort karo
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        
        List<int[]> mergedIntervals = new ArrayList<>();

        
        for (int[] current : intervals) {

            if (mergedIntervals.isEmpty() || mergedIntervals.get(mergedIntervals.size() - 1)[1] < current[0]) {
                mergedIntervals.add(current);
            } 
            else {
                int[] lastInterval = mergedIntervals.get(mergedIntervals.size() - 1);
                lastInterval[1] = Math.max(lastInterval[1], current[1]);
            }
        }

        
        return mergedIntervals.toArray(new int[mergedIntervals.size()][]);
    }
}
